api/processors/auxil/algorithms.py

```diff
@@ -0,0 +1,171 @@
+# DO NOT USE IN PROD, WORK IN PROGRESS
+
+from collections.abc import Iterable
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.utils import timezone
+
+from api.models import AgeRange, Group, Student, Teacher
+from api.models.auxil.constants import (
+    DEFAULT_LESSON_DURATION_MIN,
+    MAX_AGE_KIDS_GROUP,
+    MAX_AGE_TEEN_GROUP,
+)
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.communication_language_mode import CommunicationLanguageMode
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import (
+    GroupProjectStatus,
+    StudentSituationalStatus,
+    TeacherProjectStatus,
+    TeacherSituationalStatus,
+)
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupBuilderAlgorithm:
+    @dataclass
+    class GroupSizeRestriction:
+        # both min and max student numbers are inclusive:
+        # group size must fit into [min; max]
```

Comment:
I had a question about the rationale behind making the upper boundary inclusive, but I understand I'll just be wasting everybody's time :) So let's at least make it a docstring so one can see this in an IDE

```suggestion
        """Restriction of a group size. 
        
        Note:
            Both min and max student numbers are inclusive: group size must fit into [``min``; ``max``].
        ```
```

Reply:
I see, good question though! The only one was (probably...) that I was translating min and max restrictions from our discussions into my head and to the code without changing. Now I'm thinking about it and it's even not bad, even though not conventional (which I wasn't thinking about initially, again).

Now I think it may be more readable if we want to create "groups of 5-10 people", and hence the restriction values are 5 and 10. I think in this particular case it might create less confusion. Or maybe for some people - more confusion. I don't know, haha. Docstring is a proper idea either way!

Reply:
Well we did the same (inclusive upper bound) for age ranges for the same reason (avoiding confusion for non-programmers), so why not do it here as well. As long as it's well-documented :)